### ISSUE TITLE
Problems due to adding IPython profile dir (~/.ipython) to sys.path

### ISSUE THREAD

**Issue by ErikRtvl** [ipython/ipython#13294]
When starting IPython, the profile directory gets added to sys.path. This means any `import` statement will also look for modules in that folder and in some cases allow you to import packages that aren't installed.

For instance, after using the Cython magic, I can import cython even when cython is not installed in my environment.

This causes issues when using packages/modules that rely on ImportError exceptions to check for optional features, for instance. We have a package that uses `cython.compiled` to check if one is using a plain-code or compiled version of a package. Due to IPython's behaviour regarding the profile directory, we cannot rely on the ImportError anymore - because `import cython` will just work.

I guess IPython uses the addition to sys.path to discover extensions and such. But giving the interference with module discovery, I think different ways for this should be found.

**Comment by Carreau** [comment]
Yes, this machinery is quite old, it pre-dates pip, so we should likely find other ways to do that. I think there was some recent changes in this area, maybe the master branch behaves better now.

**Comment by ErikRtvl** [comment]
Thanks. I will give the master branch a try soon and post the outcomes here.

**Comment by ErikRtvl** [comment]
The path `C:\Users\<USERNAME>\.ipython` is still being added to `sys.path` with the master branch.

The responsible code for this is this:
https://github.com/ipython/ipython/blob/9c54f3ee28612e334f6f4b609cfc2950b1de62d2/IPython/core/application.py#L289-L311

I tried removing the "sys.path.append" statement, and run IPython (and Jupyter Notebook as well), no issues in both. Also looking through the Jupyter code, I don't have the idea that it relies on having this folder in the `sys.path`. But I did not test thoroughly.
